wreck: rework job.submit, job.create, wrexecd.run handlers with continuations

garlick opened the pull request:
This is a hopefully somewhat better structured version of PR #1471.

This one doesn't add a hash of active jobs and doesn't limt the job module's ability to ingest jobs on ranks other than 0.

It adds a `struct wjob` data structure (with unit tests), and reworks the `job.submit`, `job.create`, and `wrexecd.run` request handlers to manipulate the KVS etc using continuations  rather than synchronous RPC's, to improve job module responsiveness.

It also satisfies a long standing feature request for an ability to synchronously publish events; that is, get a response indicating that the event has received a sequence number, as discussed in #337 and #342.  That was necessary to remove the final synchronous step in job.create and job.submit, where the handlers block listening for an event to be published before sending the response to the user.

FWIW I reran a little script inspired by @trws that that submits 50K jobs (wtih the sched module check disabled), running 200 submits in parallel:
```sh
#!/bin/bash
ulimit 65536
seq 1 50000 | xargs -n 1 -P 200  \
    flux submit -n 1 sleep 0
```
Before this PR I got 134 job per second.  After I get 442.

I haven't tried to probe the effect on job launch, but it should improve latency under load - for example if there are many small jobs being launched, the overhead of checking a new job _not_ targetting the local rank should not slow down (as much) the launch of one that _does_ target the local rank, since new events can start to be handled before the last one has completed.  E.g. there could be many fetches of `R_lite` for different jobs occurring simultaneously on each rank.

Sorry the "rework" commits are basically unreadable.  They each effectively rewrite a handler and its helper functions as a chain of continuations.  It might be better to look at the final result.

grondo commented:
Very nice result @garlick! Thanks for thinking of doing that test. I wonder if we'd get even better throughput  with a bulk submit tool that can issue `job.submit` requests asynchronously? A 4x improvement is good enough for now though, and really proves that blocking rpcs should be avoided in services!

I'll just review the final result, but I'm sure everything is much improved.

Thanks also for renaming `struct job` to `struct wjob`, it was going to be one of my comments on the last PR. I would have even suggested `struct wreck_job` to reduce chance of future conflicts.

Just a thought: On commit descriptions, I've gone to using `wreck:` to prefix all of the `modules/wreck` based hacks and kludges. Not that this PR is a kludge, but there is some confusion as when to use `modules/wreck` vs `modules/job` (which may conflict with a future subsystem), whereas simply `wreck:` could easily be grepped for in history (or more likely `grep -v`).

Nice improvement on the event publishing too, I'm sure that helped throughput as well!

garlick commented:
Thanks - I'll fix the commit messages, and look at renaming `wjob` to `wreck_job`.

The test above does run `flux submit` in parallel (up to 120 of them) due to  xargs -P, so those results are probably similar to what you'd get from a single asynchronous submit tool.

dongahn commented:
> It also satisfies a long standing feature request for an ability to synchronously publish events; that is, get a response indicating that the event has received a sequence number, as discussed in #337 and #342. That was necessary to remove the final synchronous step in job.create and job.submit, where the handlers block listening for an event to be published before sending the response to the user.

@garlick, this is awesome and will make certain concurrent programming much more intuitive without hitting performance hit!

One question so that I understand its semantics. The sync eventing is suitable when you have a single receiver who needs to be sync'ed with the sender? We will have to use different mechanism if yoi have multiple subscribers?

garlick commented:
> One question so that I understand its semantics. The sync eventing is suitable when you have a single receiver who needs to be sync'ed with the sender? We will have to use different mechanism if you have multiple subscribers?

No, event delivery is still "open loop" or "fire and forget".  This RPC returns when the event has been accepted on rank 0, meaning it has a place in the event sequence.  The issue it was intended to solve, as I understand it, event ordering in the following scenario:

Say A and B both implement a service that publishes an event upon request.

If C first sends an RPC to A generating E1, then sends an RPC to B, generating E2, what order do subscribers see the events?   We don't know, it could be E1 then E2, or E2 then E1.  This is because events are "fire and forget" from A and B's perspective, and maybe A's broker is busier than B's, so its messages reach the rank 0 broker after A's.

This new thing just provides an optional way to publish events with a response from the rank 0 broker indicating that the event has been accepted and assigned a sequence number.  So if A and B wait for the response to the new RPC before responding to C, then C knows the events will be sent in the order requested.

Before in the job module, we were listening for events to loop back to the sender, which accomplished the same thing, but was a little more awkward to incorporate into the chain of continuations pattern in use in this rework.

It's kind of a niche use case I think, so don't get too excited :-)

garlick commented on src/broker/broker.c:
```diff
@@ -1276,6 +1276,46 @@ static void cmb_disconnect_cb (flux_t *h, flux_msg_handler_t *mh,
+ * User sends request message with topic "cmb.pub.<topic>" to rank 0.
+ * Service publishes event with same payload as request, topic=<topic>,
+ * then responds with success or failure.
+ * The synchronization use case driving the need need for this is
```
oops thanks!

garlick commented on src/modules/wreck/job.c:
```diff
@@ -151,97 +150,75 @@ static char * realtime_string (char *buf, size_t sz)
-        free (topic);
-        return;
+    if (snprintf (topic, sizeof (topic), "cmb.pub.wreck.state.%s", job->state)
+                                                        >= sizeof (topic)) {
```
Thanks, I'll add a comment there.

I didn't see the need for the extra JSON decode/encode.  It could certainly be done that way though!

grondo replied:
Sorry, I did guess *why* you did it that way, and it seemed smart to me! I was just wondering if this was different enough from other "RPCs" that it might confuse users of the interface. However, if the interface is hidden behind an API call or if this form of RPC topic becomes useful as standard practice, I don't see a problem with it (threw me at first though)

garlick replied:
I did add a little note about the topic string transformation just now.  Hopefully that helps?

garlick replied:
Oh sorry, more generally I was thinking that to generalize this we might want to implement some API functions for event publishing.  (Currently we encode an even and pass it to `flux_send()`):
```c
// future is fulfilled once sequence is assigned
flux_future_t *flux_event_publish_seq (flux_t *h, const char *topic, const char *json_str);

// fire and forget
int flux_event_publish (flux_t *h, const char *topic, const char *json_str);
```
Then whatever libertiy we're taking behind the scenes with topic strings is just an implementation detail where a comment will suffice.

Updating the API seems like its own PR though.   Am I still missing the point of your question?

grondo replied:
Yeah, that satisfies my curiosity.
In this case updating the API is definitely its own PR. However, in future cases where we aren't just updating throwaway code, it is much preferable IMO to add new service, add new API calls, add users of the new API calls, rather than adding API calls after the fact and having to go through and find/update all callers. (i.e. updating the API isn't *always* going to be its own PR)
(After writing that down it seems obvious, sorry!)

garlick replied:
At the time I was thinking of it as experimental and maybe not useful beyond this module.  But having written down those proposed API calls, maybe I'm convincing myself it's not a bad idea?  Anwyay, if you're OK with deferring to another PR, let's do that.  I'll open an issue to track it.

garlick commented on src/modules/wreck/job.c:
```diff
@@ -151,97 +150,75 @@ static char * realtime_string (char *buf, size_t sz)
+    if (flux_request_decode (msg, NULL, &json_str) < 0)
+        goto error;
+    if (!json_str || !(o = json_tokener_parse (json_str)))
+        goto inval;
```
Hmm, maybe.  Let me see how much work that is.

grondo replied:
Don't worry about it if it is even more than a little work. I can take a look at it in the next round.

garlick replied:
It seems somewhat orthogonal to this PR so I'll skip that for now.

garlick commented on src/modules/wreck/job.c:
```diff
@@ -270,139 +247,178 @@ static bool sched_loaded (flux_t *h)
+                                          "ncores", &job->ncores,
+                                          "walltime", &job->walltime) < 0)
+        goto error;
+    wjob_set_state (job, "submitted");
```
Good idea.

garlick replied:
Would like to discuss the constraints on these messages before doing something like this.

garlick commented on src/modules/wreck/job.c:
```diff
@@ -590,33 +605,36 @@ static bool lwj_targets_this_node (flux_t *h, const char *kvspath)
-            flux_log (h, LOG_INFO, "No %s.R_lite: %s",
-                      kvspath, flux_strerror (errno));
+            flux_log (h, LOG_INFO, "No %s: %s", key, flux_strerror (errno));
+        if (lwj_targets_this_node (h, job))
```
Yeah, I was just being lazy there - hoping that code would go away soon.  Maybe not?

grondo replied:
It didn't seem like `R_lite` producing sched was imminent, but maybe @dongahn can comment here. Hate for you to do extra work just to have it removed in a week.

garlick replied:
Just pushed a change to implement this.

garlick commented:
Pushed some minor changes:
* rename wjob to wreck_job
* fix commit wording, and wreck: prefix

(those were squashed)

Then I added a couple of small commits that began addressing review comments which can be squashed later.  More coming tomorrow.

grondo commented:
LGTM, ready for merge?

garlick commented:
Let me squash down the incremental development, then OK.

garlick commented:
Squashed.

I restarted a couple of builders that failed installing packages.

garlick commented:
Thanks!
